gradle-jdks-setup/src/main/java/com/palantir/gradle/jdks/setup/GradleJdkInstallationSetup.java

```diff
@@ -0,0 +1,109 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.jdks.setup;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.FileVisitResult;
+import java.nio.file.FileVisitor;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class responsible for installing the current JDK into {@code destinationJdkInstallationDir} and importing the
+ * certificates specified by their serialNumbers & alias name from {@code certsDir} into the JDK's truststore.
+ * A certificate will be imported iff the serial number already exists in the truststore.
+ * The class will be called by the Gradle setup script in
+ * <a href="file:../resources/gradle-jdks-setup.sh">resources/gradle-jdks-setup.sh</a>.
+ */
+public final class GradleJdkInstallationSetup {
+
+    public static void main(String[] args) throws IOException {
+        StdLogger logger = new StdLogger();
+        if (args.length != 2) {
+            throw new IllegalArgumentException("Expected two arguments: destinationJdkInstallationDir and certsDir");
+        }
+        Path destinationJdkInstallationDir = Path.of(args[0]);
+        Path certsDir = Path.of(args[1]);
+        atomicCopyJdkInstallationDirectory(logger, destinationJdkInstallationDir);
+        Map<String, String> certNamesToSerialNumbers = extractCertsSerialNumbers(logger, certsDir);
+        CaResources.maybeImportCertsInJdk(logger, destinationJdkInstallationDir, certNamesToSerialNumbers);
+    }
+
+    private static void atomicCopyJdkInstallationDirectory(ILogger logger, Path destinationJdkInstallationDirectory)
+            throws IOException {
+        Path currentJavaHome = Path.of(System.getProperty("java.home"));
+        Path jdksInstallationDirectory = destinationJdkInstallationDirectory.getParent();
+        if (!jdksInstallationDirectory.toFile().exists()) {
+            Files.createDirectories(jdksInstallationDirectory);
+        }
+        Path lockFile = jdksInstallationDirectory.resolve(destinationJdkInstallationDirectory.getFileName() + ".lock");
+        try (FileChannel channel = FileChannel.open(lockFile, StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
+            channel.lock();
+            // double-check, now that we hold the lock
+            if (Files.exists(destinationJdkInstallationDirectory)) {
+                logger.log(String.format("Distribution URL %s already exists", destinationJdkInstallationDirectory));
+                return;
+            }
+            logger.log(
+                    String.format("Copying JDK from %s into %s", currentJavaHome, destinationJdkInstallationDirectory));
+            FileUtils.copyDirectory(currentJavaHome, destinationJdkInstallationDirectory);
```

Comment:
Even thought we have the locks, we should probably still copy to another directory (perhaps the target path with some UUID appended?) then do an atomic move.

Reply:
Did we try just doing a move here? It might just work (and save some time).

Reply:
> Did we try just doing a move here? It might just work (and save some time).

It seems to work - the tests pass: https://github.com/palantir/gradle-jdks/compare/cr/gradle-jdk-script...cr/check-move-while-running?expand=1
All the files are moved correctly to the new destination. After the move there are still some left-over directories ~90 eg. 
`<inProgressDir>/amazon-corretto-11.0.21.9.1-linux-aarch64/legal/jdk*`, `<inProgressDir>/amazon-corretto-11.0.21.9.1-linux-aarch64/lib/security` Probably the process was keeping a handle on these ? But in that case I don't really understand though why `<inProgressDir>/amazon-corretto-11.0.21.9.1-linux-aarch64/bin/java` was not still present.

Reply:
Hmm, the left over files seem dodgy - I bet we can't delete them from the java code (although probably could from the shell script). Might be best to just copy then, especially with the windows issue too.